service/builder/receivers_builder.go

```diff
@@ -248,32 +248,46 @@ func (rb *ReceiversBuilder) buildReceiver(config configmodels.Receiver) (*builtR
 	return rcv, nil
 }
 
-func buildFanoutTraceConsumer(pipelineFrontProcessors []*builtProcessor) consumer.TraceConsumer {
+func buildFanoutTraceConsumer(pipelines []*builtPipeline) consumer.TraceConsumer {
 	// Optimize for the case when there is only one processor, no need to create junction point.
-	if len(pipelineFrontProcessors) == 1 {
-		return pipelineFrontProcessors[0].tc
+	if len(pipelines) == 1 {
+		return pipelines[0].firstTC
 	}
 
 	var pipelineConsumers []consumer.TraceConsumer
-	for _, builtProc := range pipelineFrontProcessors {
-		pipelineConsumers = append(pipelineConsumers, builtProc.tc)
+	anyPipelineMutatesData := false
+	for _, pipeline := range pipelines {
+		pipelineConsumers = append(pipelineConsumers, pipeline.firstTC)
+		anyPipelineMutatesData = anyPipelineMutatesData || pipeline.MutatesConsumedData
 	}
 
 	// Create a junction point that fans out to all pipelines.
+	if anyPipelineMutatesData {
+		// If any pipeline mutates data use a cloning fan out connector
+		// so that it is safe to modify fanned out data.
+		return processor.NewTraceCloningFanOutConnector(pipelineConsumers)
+	}
 	return processor.NewTraceFanOutConnector(pipelineConsumers)
 }
 
-func buildFanoutMetricConsumer(pipelineFrontProcessors []*builtProcessor) consumer.MetricsConsumer {
+func buildFanoutMetricConsumer(pipelines []*builtPipeline) consumer.MetricsConsumer {
 	// Optimize for the case when there is only one processor, no need to create junction point.
-	if len(pipelineFrontProcessors) == 1 {
-		return pipelineFrontProcessors[0].mc
+	if len(pipelines) == 1 {
+		return pipelines[0].firstMC
 	}
 
 	var pipelineConsumers []consumer.MetricsConsumer
-	for _, builtProc := range pipelineFrontProcessors {
-		pipelineConsumers = append(pipelineConsumers, builtProc.mc)
+	anyPipelineMutatesData := false
+	for _, pipeline := range pipelines {
+		pipelineConsumers = append(pipelineConsumers, pipeline.firstMC)
+		anyPipelineMutatesData = anyPipelineMutatesData || pipeline.MutatesConsumedData
```

Comment:
doesn't the need to clone data only applies for pipelines with the same receivers?

Reply:
Yes, this function is called for pipelines attached to one receiver.

Reply:
ahh i got lost in builders - i thought we were in a top level builder - 👍 this looks good